### PR TITLE
DROOLS-2815: [DMN Designer] Make Expression Type dependent on Diagram Node

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-2815

@jomarko Please consider the comment [here](https://issues.jboss.org/browse/DROOLS-2815?focusedCommentId=13621953&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13621953).

### CODE REVIEW

**Comment by manstis** [comment]
I propose this is reviewed after the PRs for the "in cell" Data Type selector PRs are all merged.

There might be some merge conflicts when this comes to be merged after those.

**Comment by manstis** [comment]
Jenkins please retest this.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNEditBusinessKnowledgeModelToolboxActionTest.java [hunk 1]
```diff
@@ -125,11 +125,23 @@ public void testAction() {
...
+        assertFalse(hasExpression.isClearSupported());
         assertEquals(bkm,
                      editExprEvent.getHasName().get());
         assertEquals(session,
                      editExprEvent.getSession());
+
+        try {
+            hasExpression.setExpression(new DecisionTable());
+        } catch (UnsupportedOperationException use) {
+            //This is expected
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNEditBusinessKnowledgeModelToolboxActionTest.java]: Could we use something like the following example in this test?

```java
@Test(expected = UnsupportedOperationException.class) 
```

**Comment by manstis** [comment]
@jomarko Updates made (in new, third, commit).

**Comment by manstis** [comment]
@karreiro @jomarko This PR is good for review.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionContainerGrid.java [hunk 2]
```diff
@@ -206,12 +211,16 @@ public void setName(final Name name) {
...
+        final List<ListSelectorItem> items = new ArrayList<>();
+        if (hasExpression.isClearSupported()) {
+            items.add(ListSelectorTextItem.build(translationService.format(DMNEditorConstants.ExpressionEditor_Clear),
+                                                 true,
+                                                 () -> {
+                                                     cellEditorControls.hide();
+                                                     clearExpressionType();
+                                                 }));
+        }
+        return items;
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionContainerGrid.java]: Just due to previous version, thinking if this is more effective:
```java
if (hasExpression.isClearSupported()) {
            return Collections.singletonList(ListSelectorTextItem
                                              .build(translationService.format(DMNEditorConstants.ExpressionEditor_Clear),
                                                     true,
                                                     () -> {
                                                         cellEditorControls.hide();
                                                         clearExpressionType();
                                                     }));
        }
        return Collections.emptyList();
```

let me know what do you think please.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNEditBusinessKnowledgeModelToolboxActionTest.java [hunk 3]
```diff
@@ -125,11 +125,31 @@ public void testAction() {
...
+
+        final ArgumentCaptor<EditExpressionEvent> eventCaptor = ArgumentCaptor.forClass(EditExpressionEvent.class);
+        verify(editExpressionEvent,
+               times(1)).fire(eventCaptor.capture());
+
+        final EditExpressionEvent editExprEvent = eventCaptor.getValue();
+        final HasExpression hasExpression = editExprEvent.getHasExpression();
+
+        hasExpression.setExpression(new DecisionTable());
+    }
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNEditBusinessKnowledgeModelToolboxActionTest.java]: Please use assertj:
```java
@Test
    public void testActionSetExpression() {
        tested.onMouseClick(canvasHandler,
                            E_UUID,
                            mouseClickEvent);

        final ArgumentCaptor<EditExpressionEvent> eventCaptor = ArgumentCaptor.forClass(EditExpressionEvent.class);
        verify(editExpressionEvent,
               times(1)).fire(eventCaptor.capture());

        final EditExpressionEvent editExprEvent = eventCaptor.getValue();
        final HasExpression hasExpression = editExprEvent.getHasExpression();

        Assertions.assertThatThrownBy(() -> hasExpression.setExpression(new DecisionTable()))
                .isInstanceOf(UnsupportedOperationException.class)
                .hasMessage("It is not possible to set the EncapsulatedLogic of a BusinssKnowledgeModel.");
    }
```

**Comment by jomarko** [comment]
@manstis manual check passed, please have a look on comments to code.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionContainerGrid.java (hunk 2 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionContainerGrid.java]: Updated.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNEditBusinessKnowledgeModelToolboxActionTest.java (hunk 3 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/canvas/controls/toolbox/DMNEditBusinessKnowledgeModelToolboxActionTest.java]: Updated.

**Pull request merged**